Parser refactor

gdolle opened the pull request:
Major refactoring for the parser. Breaking changes (3.1.0 ?)

Changelog

- Fix options bugs (random behaviour between book.json and local command, ace plugin does not support quotes in command options)
- Add Url support #42 (sync http request using [sync-request](https://www.npmjs.com/package/sync-request) modules + check url [valid-url](https://www.npmjs.com/package/valid-url))
- Refactor options using immutable maps (more robust). It makes easier future new options, see `src/options.js`
- Support real types (not retrocompatible for book options). For example `edit:true` instead of `edit:"true"`,  to be more consistent (this fix an ace plugin bug).
- Add parser checks for types.
- Add different warn level.
- `template`, `fixlang`, `lang`, `theme` are now also local/global options. We can set a template per code, a theme per code, etc.. or globally.
- Refactor test for these last changes.
- Add new tests for URL
- End support of `lang-typescript` to use uniform options `lang:"typescript"` to be consistent with other options and to simplify functions.
- Support multiple snippet codes. Now we can do `[import:"tag0,tag1,..."](fixtures/test.js)` to import several snippets into one codeblock , for example
```cpp
//! [tag0]
int a;
//! [tag0]
// A comment or a code
//! [tag1]
int b;
//! [tag1]
```
will produce
```cpp
int a;
int b;
```

It should work far better with this PR :)

I've been noticed that ace plugin might crash after a page reload in previous version, but I couldn't reproduce that problem. However, these changes fix some bugs and "might" be fix for that behaviour, we will see.

NB: It might be interesting to have a look to async operation in the future to optimize the code.

@azu waiting for your review, you'll have some work ;)

gdolle committed:
rename and add new examples per template

gdolle committed:
Remove unrequired plugin

Not required in fact.

`gitbook-plugin-ace` is required only if user pass "ace" or "acefull" to the template
option. It is specified in the README.md.

gdolle committed:
Update readme

gdolle committed:
Add check plugin load order

- Add winston-color for color logs
- produce a warning if gitbook-plugin-ace is not loaded (ace template won't work)
- produce an error if gitbook-plugin-ace is loaded but before include-codeblock

gdolle committed:
update log message

gdolle committed:
update travis tests

gdolle committed:
Fix Readme

gdolle committed:
use path package

gdolle committed:
separate ace check

gdolle committed:
Improve example

gdolle committed:
Improve custom example

gdolle committed:
Improve default example

gdolle committed:
Refactor parser

- Should be cleaner and less error prone
- Define global book option map, and possible key-value map (same as in
  readme).
- Use immutable Map to avoid unwanted changes.
- Pass `template`, `unindent` function arguments into the key-value map
  directly for consistency.
- Systemize key-value argument for each functions.

gdolle committed:
Add inline code remark

- include-codeblock does not work if we use it inline.
for example:

```
blabla [import](fixtures/test.js) blablabla
```

NB: we might have some works to add newline before and after block.

gdolle committed:
Update tests

gdolle committed:
Remove log

gdolle committed:
up package version

gdolle committed:
Add multi snippets feature

gdolle committed:
Update test using templates from map

gdolle committed:
Add support equal char for

gdolle committed:
Fix template quotes + up version 3.1.0 major

- It seems ace does work with boolean instead of strings for booleans
type keys.
- Update version 3.1.0, changes will be major

gdolle committed:
Update readme

gdolle committed:
Rename options to avoid confusion

gdolle committed:
Separate options maps

Options can be set globally or locally:

- Set an immutable default map book.json global plugin options
- Set an immutable default key-value map for local command options.
- Add functions to check map types

gdolle committed:
Clean option file

gdolle committed:
Move convertValue

gdolle committed:
Refactor parser

- Refactor parseVariableFromCommand regex to be more comprehensible.
- Handle boolean type
- Add parse check depending on the value type and convert the parsed
  value to its type (JSON.parse)
- Fix type check. Default l/g options map should not have `undefined`
  value as value type is deduced from the map.

gdolle committed:
Update marker check with empty string

- Update check for empty string which is not valid.

gdolle committed:
Update package options

- Add default value options in package.json directly. Options map are
  updated from it.
- Fix package default type to be consistent (avoid string). Now boolean
  is `true` and not `"true"`.

gdolle committed:
Revert options get

gdolle committed:
Update + new test

- Update parse tests to handle better tests (simpler & different case).
- Add test for multiple markers

gdolle committed:
merge from upstream

gdolle committed:
Merge branch 'master' into parser-refactor

gdolle committed:
Refactor lang key

break backward compatibility

- set hardcode lang as book option, or command option.
- Remove `lang-javascript` and use standard key-val format
`lang:"javascript"` or `lang:".js"`. If lang or extension (via lang) is not set,
 then lang is determined rom file extension.
- Create template.js file for template specific codes.

gdolle committed:
Add support of URL (sync)

issue #42

- Perform http GET request and retrieve url body.
- Asynchronous http (module sync-request) with caching, gzip
- Redirection not allowed.

NB: Future optimization could be to provide async operation for the
plugin.

gdolle committed:
Update readme

gdolle committed:
Update readme

azu commented:
Oh, large diff.. 
I'll look it lator.
Can you separate refactor PR and feature PR if possible.

gdolle commented on src/parser.js:
```diff
@@ -36,65 +42,127 @@ export function splitLabelToCommands(label = "") {
+ * @param {string} originalPath
+ * @return {string}
+ */
+export function getContent( filePath, originalPath )
```
Part of feature #42

gdolle commented on templates/ace-template.hbs:
```diff
@@ -1,5 +1,5 @@
 {% if file.type=="asciidoc" %}++++{% endif %}
-{%ace edit="{{edit}}", check="{{check}}", theme="{{theme}}", lang="{{lang}}" %}
+{%ace edit={{edit}}, check={{check}}, theme="{{theme}}", lang="{{lang}}" %}
```
Major bug fix here

gdolle commented:
@azu Actually, features are minors and deeply dependent of the of refactor. It'd be quite difficult to separate now. 

Would it be ok to keep it as one PR ?

The related features changes are
- multiple markers (parseVariableFromLabel => change the regex to have several comma ~1line)
- URL parser ( getContent => load url or file ~5lines) 
+ the 1 tests for each in `test-parser.js`

(I added comments on file changes to help)

azu commented:
@gdolle fmm, I want to separate big PR to smll PRs by following reasons.

> Add Url support #42 (sync http request using sync-request modules + check url valid-url)

Sync request is simple solution, but it is not correct solution.
Because GitBook support async hook.

- https://toolchain.gitbook.com/plugins/hooks.html

(Not actuall code)
```js
// THIS IS IMAGINE
module.exports = {
    hooks: {
        "page:before": function(page) {
            var options = this.options.pluginsConfig["include-codeblock"];
            var pageDir = path.dirname(page.rawPath);
            var parseResults = parse(page.content, pageDir, options);
            var promises = results.map(result => {
                return requestUrlAndCreateReplaceContent(result).then(newContent => {
                     replaceContent(page, newContent);
                });
            });
            return Promise.all(promises); // GitBook support async by returing promise
        }
    }
};
```

and

[Immutable.js](https://facebook.github.io/immutable-js/ "Immutable") is overlarge for this project.
I recommened that use ES2015 [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map "Map") + Poilyfill insteadof immutable.js.

-----

Of course, Welcome to [bugfix](https://github.com/azu/gitbook-plugin-include-codeblock/pull/43#discussion_r97314114), but the bugfix should not contains other feature.
I hesitate to merge the bugfix PR.

Sorry, slow response.

gdolle commented:
Does ES2015 Map guaranty const Map values ?

Actually refactor parser:
1. immutable Map for read only (more robust to my opinion) 
2. handle key-val types (boolean, string, number). The "bug fix" is a consequence of the parser refactor (maybe my single comment is error prone) it's not really a feature.

I agree for the url feature, it is from far not perfect. I'll separate it in another PR.

azu commented on src/parser.js:
```diff
@@ -36,65 +42,127 @@ export function splitLabelToCommands(label = "") {
+    const spaces_ex = "\\s*";
+    const quotes_ex = "[\"']";
+
+    Object.keys(kv).forEach(key => {
```
Can ES2015 Map does same thing?

```js
const resultMap = new Map();
kvMap.forEach(key => { 
   // something
   resultMap.set(key, foo);
}
return resultMap;
```

azu commented on src/parser.js:
```diff
@@ -103,114 +171,69 @@ export function embedCode({lang, filePath, originalPath, label, template, uninde
+    const kv = kvMap.toObject();
+    const count = hasTitle(kv) ? codeCounter() : -1;
+    checkMapTypes( kvMap, "generatedEmbedCode" );
+    const contextMap = kvMap.concat( immutable.Map({
```
`var mergedMap = new Map([...kvMap, ...otherMap])`

```js
var a = new Map([["k1", "v1"], ["k2", "v2"]]);
var b = new Map([["k3", "v3"], ["k4", "v4"]]);
var mergedMap = new Map([...a, ...b]);
console.log(mergedMap); // Map { k1: "v1", k2: "v2", k3: "v3", k4: "v4" }
```


```js
// Object to Map
new Map(Object.entries({ k : 1 }))
```

azu reviewed (commented):
> Does ES2015 Map guaranty const Map values ?

Yes and No.

Immutable.js is great library, but it is not first class in JavaScript.
In other word, When other constributor see the code that includes immutable.js, they have to learn about immutable.js. I think that it contains cost.

I've written some Map examples.

If you actualy want to use immutable.js, I can't prevent it.

> I agree for the url feature, it is from far not perfect. I'll separate it in another PR.

Ok.

azu commented:
(Currently, this project mixed codeing style, We should apply ESLint  #39)

gdolle committed:
Remove immutable.js

- Remove immutable package
- Remove Map object to use frozen Object instead (guaranty unmodified
  default map object).
- Update tests.

gdolle committed:
Remove parser feature

- Remove URL parser source code (different branch + need update async)
- Remove URL parser test
- Remove package.json dependencies

gdolle commented:
@azu I updated the code. 
- Remove immutable: I used object instead of Map and froze object that should be immutable. It should be understandable for future contributors.
- Remove URL feature.

I didn't apply yet Eslint, I'll have a look into that.

gdolle committed:
Add Eslint config #39 + fix coding style

See issue #39

- Use this convention https://github.com/textlint/textlint/blob/master/.eslintrc.js
- Change Error to Warn for unused vars and allow console logs.
- Add new package script : check => run eslint for javascript files in src/ and test/.
  (Do not add test/fixtures, test/patterns, ...).
- Add travis eslint check.

gdolle committed:
add eslint config

gdolle commented:
@azu I added eslint config following #39 and fix current code following the convention.
I also added a "check" script which test eslint on src/ and test/.
This script is executed in travis, therefore we can see if something is not good with respect to the coding style.
Let me know if it's ok for you like that. ;)

azu commented on src/parser.js:
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
+  // inspired from https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/string/strip.rb
+  if((s === undefined) || (s === "")) {
+return s;
+}
```
Can you format indent or Run `$(npm bin)/eslint --fix src/ test/`?

gdolle replied:
weird eslint does not fix this lines..

azu commented on src/parser.js:
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
+                break;
+            default:
+                logger.error("include-codeblock: parseVariablesFromLabel: key type `" +
+                    typeof defaultKeyValueMap[key] + "` unknown (see options.js)");
```
add `breaK;` to `default:` block.

azu commented on src/parser.js:
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
     }
-    if (unindent || keyValueObject.unindent) {
-      content = strip(content);
+    if (unindent == true) {
```
use `===` insteadof `==`.

OR

```js
const unindent = kvm.unindent !== undefined ? kvm.unindent : false;
...
if (unindent) { ... }
...
```

azu commented on package.json:
```diff
@@ -11,7 +11,7 @@
     "url": "https://github.com/azu/gitbook-plugin-include-codeblock/issues"
   },
-  "version": "3.0.1",
+  "version": "3.1.0",
```
PR is not include bump version.

azu commented on package.json:
```diff
@@ -24,6 +24,7 @@
   "scripts": {
     "build": "NODE_ENV=production babel src --out-dir lib --source-maps",
+    "check": "eslint src/*.js test/*.js",
     "watch": "babel src --out-dir lib --watch --source-maps",
```
Maybe, `lint` command is meaningful.

Also, To define `"lint:fix": "eslint --fix src/*.js test/*.js"` is useful.

azu reviewed (changes requested):
Thanks.
I've leave some comments

gdolle committed:
Update reviewed code

gdolle commented:
Sorry for the late reply. I updated the code relatively to your last review

azu commented:
@gdolle Thanks for great work!
